src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/explorer/TracingMessageConsumer.java

```diff
@@ -0,0 +1,214 @@
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ ******************************************************************************/
+package com.redhat.devtools.intellij.quarkus.lsp4ij.console.explorer;
+
+import com.redhat.devtools.intellij.quarkus.lsp4ij.settings.ServerTrace;
+import org.eclipse.lsp4j.jsonrpc.JsonRpcException;
+import org.eclipse.lsp4j.jsonrpc.MessageConsumer;
+import org.eclipse.lsp4j.jsonrpc.MessageIssueException;
+import org.eclipse.lsp4j.jsonrpc.RemoteEndpoint;
+import org.eclipse.lsp4j.jsonrpc.json.MessageJsonHandler;
+import org.eclipse.lsp4j.jsonrpc.json.StreamMessageConsumer;
+import org.eclipse.lsp4j.jsonrpc.messages.Message;
+import org.eclipse.lsp4j.jsonrpc.messages.NotificationMessage;
+import org.eclipse.lsp4j.jsonrpc.messages.RequestMessage;
+import org.eclipse.lsp4j.jsonrpc.messages.ResponseMessage;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.time.format.DateTimeFormatter;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Logger;
+
+import static java.util.logging.Level.WARNING;
+
+/**
+ * A {@link MessageConsumer} that outputs logs in a format that can be parsed by the LSP Inspector.
+ * https://microsoft.github.io/language-server-protocol/inspector/
+ */
+public class TracingMessageConsumer {
+    private final Map<String, RequestMetadata> sentRequests;
+    private final Map<String, RequestMetadata> receivedRequests;
+    private final Clock clock;
+    private final DateTimeFormatter dateTimeFormatter;
+
+    public TracingMessageConsumer() {
+        this.sentRequests = new ConcurrentHashMap<>();
+        this.receivedRequests = new ConcurrentHashMap<>();
+        this.clock = Clock.systemDefaultZone();
+        this.dateTimeFormatter = DateTimeFormatter.ofPattern("KK:mm:ss a").withZone(clock.getZone());
+    }
+
+    /**
+     * Constructs a log string for a given {@link Message}. The type of the {@link MessageConsumer}
+     * determines if we're sending or receiving a message. The type of the @{link Message} determines
+     * if it is a request, response, or notification.
+     */
+    public String log(Message message, MessageConsumer messageConsumer, ServerTrace serverTrace) throws MessageIssueException, JsonRpcException {
+        final Instant now = clock.instant();
+        final String date = dateTimeFormatter.format(now);
+
+        if (messageConsumer instanceof StreamMessageConsumer) {
+            return consumeMessageSending(message, now, date, serverTrace);
+        } else if (messageConsumer instanceof RemoteEndpoint) {
+            return consumeMessageReceiving(message, now, date, serverTrace);
+        } else {
+            return String.format("Unknown MessageConsumer type: %s", messageConsumer);
+        }
+    }
+
+    private String consumeMessageSending(Message message, Instant now, String date, ServerTrace serverTrace) {

```

Comment:
Maybe the different formats can be stored as private static final constants

Reply:
I would like to keep this code to keep the maximum similarity from the origin code https://github.com/eclipse-lsp4j/lsp4j/blob/main/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java